### ISSUE TITLE
Define a Cluster of Nodes at a specific position (North/West/East/South)

### ISSUE THREAD

**Issue by BjBe82** [eclipse/elk#854]
I try to cluster a set of Data in different areas of the diagram. Let’s assume there are the following Elements:
- RIU
- BAT_1
- BAT_2
- PCDU
- Others

The goal is to have RIU in the TopLeft, BAT_1/BAT_2 in the BottomLeft, PCDU in the TopRight and Others in the BottomRight of the diagram.

To cluster the elements a hierarchical node will be added representing the orientation (maybe not required). I tried to use "position" in combination with "partition" to force the grouping node in the corresponding direction.
- TopLeft -> partition:1 / position:0,0
- BottomLeft-> partition:1 / position:0,100
- TopRight -> partition:1 / position:0,0
- TopRight -> partition:1 / position:0,100

[Example Model](http://rtsys.informatik.uni-kiel.de/elklive/elkgraph.html?compressedContent=IYGw5g9gTglgLgCwLYC4AEJgE8CmUcAmAUDiANYB0BM+AxnDBAHboBKAkgOIASAKkUwgEcAGWAAjUgGcKAB0y0cSHEzjoARAG1uAfQDCAUQByvA6wA0aAGo7eAeQAKl9kYDK7ACIGAuuoFCcVxgALxwKWmYpOChgGFUpDU0jOy8dEQBBACEDEVdfIllgKAYGZjiwCmB6GAA3YDgcdGiAVxwiBBg8ItoELG5gJgIQcvQXPREAVVS9bnYRD1ZjIlooCCkpcoBZOJgkEPrGJgopJRh2VS7qmsa0Frb-YTQmAEZntABvIjRvjGwIZrgaE0aA2oXQzwADBDLJCIWhvF8foVivBDuU5EUSodwYjvgA9WRrVHMDQACghFGhaApEIAlH4fk8AmJJCAZPIqkoVGo0FpdIYTGZLDZ7E40C53F58ozMKzebwILIRDgAGZwdQfX5Yf6A4Ggm4AVipzwAbPC0ABfIi4pmPDgTD42xmCYQs6RyBRc1SJfnGUwWaz6P1C8VuTw+BmMn6y0i8+0a95anVAkEhG4AJgALDCzd5LTarVaXTgns9047GcisWUmBUq8SWGhnjaCUTSo31OTKTCoZT6Tbi262R7OcpvbztEHBQGRY5nGGpZHvjGQLzMhA4HAIEhlWqE0mASn9egjTnzVaBwE0Jl0rwdG9PlGfsWgqFwpForF4olkqkMtlcmlJ9bVECR3Q5RQxx5Pkp39YVYJDCVwyAp8VzXW9733TBtUPPU03QE1jVzfNGStZ0rxvO9y0fYDyOEV8wgiJgohiOI4ASCdfwMNIshyPIlyjQcwOHCCvWgycBTgwNJMQhcIydaNhPQqisL+XDUzBNBCLPPMyJ+QtrWLJ50wfG163bdFzOxNB0xbQkNnbMkaUsGl+zo0DWXZT0oJ9BCZ1sOdQ0leSZSU9QFVkVgYDABB1U1bDkzwzTTybYiL3ctAHD0DwHRop8XzTd9mM-NiOK0LieIA-iFO+ITPJHSDuV8mT-Ja+dgpQqM0PULKctUnDdQ0w0iPPAsiCLK8mHTaizMxBtLLmxybLsttrM7Zym17OkBLq8DvKaidfWneDRXa5CBO69dN23KKYrixMEvU480BS01RsvR47EQPApArYCCrfJiWK-difxSbj-z4zr3KHLzRwOmC2uk4MAyQxcat+OV1C+hAfv6xKhpPEbdLGi8SAIMASym54KF6h0AFoAD5S3TChKPvb5ycp4yabptAmeM1mcZ+znCG5l4aftfnmamoXvqgKQuap14KClgWXlZ9n0yAA)

But as it looks like the algorithm just adds new layers to the partitions which causes a not intended behavior. I also tried to force one layer per partition by using “layerConstraint” but using “layerConstraint” FIRST for TopLeft and BottomLeft causes an exception: 
“UnsupportedConfigurationException: Node 'TopRight' has its layer constraint set to FIRST_SEPARATE, but has at least one incoming edge”

Is there a other way to configure the intended behavior?

**Comment by soerendomroes** [comment]
You can't have an edge from TopLeft to BottomLeft, since it is an in-layer edge. Can you draw our desired drawing as you would like it to be?

**Comment by BjBe82** [comment]
TopLeft, BottomLeft, TopRight and BottomRight are only used to group elements in that area like a cluster. Then i tried to position the “cluster” nodes in a way that they are top-left, bottom-left, top-right and bottom-right oriented in the Diagram. 

I tied to achieve something like this:


![grafik](https://user-images.githubusercontent.com/61700870/179251515-eeaf7a1b-3a5d-4a01-84cb-33e4f51ea876.png)

**Comment by soerendomroes** [comment]
This does not work since you introduce in-layer edges. What you need is for example box or rectpacking layout for your grouping elements and maybe layered at the bottom. But I guess the edge routing does not handle this very well, even if you introduce ports for all outgoing edges (which you have since only layered can do hierarchy-aware layout).
To create a drawing as you describe is not trivial and is to my knowledge not supported by ELK. However, feel free to try out the pointers I gave you.

**Comment by BjBe82** [comment]
Thx for the answer and hints so far.

Tested: rectpacking + layered algorithm
[Example Model](http://rtsys.informatik.uni-kiel.de/elklive/elkgraph.html?compressedContent=IYGw5g9gTglgLgCwLYC4AEUCmBjOAHYbAaxgDswAoUiAE0wBlgAjTEAZwDo8RDMlNScdACIA2gAkA+gGEAogDkAKrIBKAGjQA1SYoDyABQ0BJeQGUjAEVkBdYVVqZTMAF6YO2CKTZwowMnDYRUXldK0l6AEEAIVl6Uw0AWRMjBIBVBMlzAC0bOwp7OjQ4CDx6TAAzODQAbwo0erRQSFhEVDQeAE9MLBo6htYiDmo6J1cOJDIYJABXVDEARgAGRbUlxds0AHpN6jQK8pw4PvrhhmZWTm5efkEgqTklVQ1tPUM0E3MrW2P285A0YSKEplSrCGq-DoQaZVURoNguTDoACsKzQ8wAbGhrGgAL75BpoU5oFRGVI1H4E06MFjsLg8bB8ARCAESGQKZTqLRsx6cj6WXIUho8GkAkmpMHVCFQmFwhHoABMABYNBisbifniGts9jQwJhiaS0ABaAB8RWBFSOeIK+uKeBUMDACCqtQJTWg8GQ6E63UwvUpDmpFzp1yZd25HOeOgMxjM-O+BOFrABQPtjudEql0LQsPhrmRqNV2OtAcK+mkFjJroJDVOozcHi8Pj8gkCLJCYUiMTiCZrJ0Df0u9MZtxZ93ZTy5D0j7zjXzsffqSf+wnLlcznWlOdl+bQKJVmOLGoo1qJTAgcGKSBBLp+Aw4WFwBGIZDAHDI8kwAHcVBAvygfGmTAfndFovQhX1-VrAcaSHUNRzEcceSjV5Y0+AVEz+AEogvK8bw3YBIWzXM5T3QtD3VH4iSiCJFEkeZyUXQkHHrdxPG8Xx-DbMQO1kcJoliUxez7KlBxDBkbmZRCI0nbRp0nPl50FJcsOEGi6PmAiiJlPNETQdFyLVTUGmM-tCnUyR5UYpjoJGBE2KbTjWyCXj+O7ISF0XUTYPEkcpNZeTOTkideTnDDF2XbDaMsrStxI3cDIPIzj2tG00HPS8ICQB0nVvN1wA9VpvUIyCqJg4MrgksMxxkoLozeRTwpUkU1NwrKcozcFN2Inc9P3NEKJLWz9V0RBujYayRJY+zGw4lsAhc0I+K7QThJrbyKuHSTw0CqNdtndC1swlrRoQcbYp63SCySo8CTxVKdT1c001yjg1zJU10ra69LQ4CyGLQOo-Seu0OrgN6Kw+s0MqvMGOFO8b6iB3VbQtSoODFY1oe+uGEagNhkZBtHwcxz6Yaym8-ui+V8goIA)

The position of the nodes looks fine but as you already mentioned the edge routing does not handle this very well. It is also not possible to introduce segmented edges (i.e RIU-> Others splitted into "RIU->TopLeft + TopLeft->BottomRight + BottomRight->Others) Trying to add an edge from RIU to TopLeft causes an exception: 
`UnsupportedGraphException: The source or the target of edge ElkEdge ElkNode "RIU" (0.0,0.0 | 30.0,30.0) -> ElkNode "topLeft" (0.0,0.0 | 30.0,30.0) could not be found. This usually happens when an edge connects a node laid out by ELK Layered to a node in another level of hierarchy laid out by either another instance of ELK Layered or another layout algorithm alltogether. The former can be solved by setting the hierarchyHandling option to INCLUDE_CHILDREN.`

**Comment by soerendomroes** [comment]
Yes, because you cannot do hierarchy-aware edges in rectpacking. What I meant was introduce ports that have the correct position (and a very small size) based on the hierarchical edges you want to introduce and only create edges in the current hierarchy level, while the ports serve as a link between the hierarchy levels.

**Comment by BjBe82** [comment]
Yes but even if i change it to edges on the same level with dummy ports the edge is not nicely routed:

![grafik](https://user-images.githubusercontent.com/61700870/183890563-953220e1-79c9-4c00-97f8-49e4a27c928b.png)

[Example Model](http://rtsys.informatik.uni-kiel.de/elklive/elkgraph.html?compressedContent=IYGw5g9gTglgLgCwLYC4AEUCmBjOAHYbAaxgDswAoUiAE0wBlgAjTEAZwDo8RDMlNScdACIA2gAkA+gGEAogDkAKrIBKAGjQA1SYoDyABQ0BJeQGUjAEVkBdYVVqZTMAF6YO2CKTZwowMnDYRUXldK0l6AEEAIVl6Uw0AWRMjBIBVBMlzAC0bOwp7OjQ4CDx6TAAzODQAbwo0erRQSFhEVDQeAE9MLBo6htYiDmo6J1cOJDIYJABXVDEARgAGRbUlxds0AHpN6jQK8pw4PvrhhmZWTm5efkEgqTklVQ1tPUM0E3MrW2P285A0YSKEplSrCGq-DoQaZVURoNguTDoACsKzQ8wAbGhrGgAL75BpoPDQKo0WZIDr4hqnNAqIypGo-AmnRgsdhcHjYPgCIQAiQyBTKdRafmPIUfSy5RkNHisgG01Jg6oQqEwuEI9AAJgALBoMVjcZSCfUiVASWSKT88QTMDQwJgaXSOKSkOS0ABaAB8RWBFTgTvNFDxBXtxTwKhgYAQVVqBKa0HgyHQnW6Np+zL+lw5XNuvPuAqewtexjMEu+BJlrABQLDEajiuV0LQsPhrmRqL12KDTIcaH00gs9JjRpODlGbg8Xh8fkEgV5ITCkRicTLw7Q6dZmeu3LuIsFz13BfFXzsq9+suEfYH9c6KqbatbaBRusxnctgfy1KYEDgxSQIOjPwDBwWC4AQxBkGAHBkPImAAO4qBAsEoD40yYD8cYtImEIpr03Z0CyFzsluOZiHmor7kW7wlsePwVv8whRN+v7-tewCQo2zbqo+7YvgaPwmmaLoWnh9pRBEiiSPMDKnqcY7uJ43i+P4s5iPOsjhNEsSmCuw7roRVycjcPKkQeQraA8e5UZ8kqnnRAJiRJ8ysexqotoiaDojx+pWg0PkjoUDmSBq0mnlSo4IvJk5KTOQRqRpS7aSeq56WyBnZsZfIWQW5n5mK1E2audkMeJQXObenEPp5z7eYaRoCWgzrkm+gG2vaX4-hAf6+v6Qnul6gUaj1TVBsGaDtb+4aRgBsbgPGrRJmxOFpg4BGpVmRk7llZk6AYxbWTpZ6VgxTGdZNdbgjeHH3u5T5orxXZhYUuiIN0bAhbp4VjBOinTgEsWhOpi5aQdImrZuhnbrmpn7lte2lklw5Fc9CCvWVV1uW21WvgSeJdjadreqU3WNR0fVjSdXWVENFoUEAA)

**Comment by soerendomroes** [comment]
You are right, rectpacking does not support port constraint or hierarchy-aware layout. Implementing this seems to be complicated, so it might take a while to add this new feature if we have the time to do this.

**Comment by soerendomroes** [comment]
Have you tried something like [this](https://rtsys.informatik.uni-kiel.de/elklive/elkgraph.html?compressedContent=IYGw5g9gTglgLgCwLYC4AEJgE8CmUcAmAUDiANYB0BM+AxnDBAHboBKAkgOIASAKkUSYQCONEwCMaAN5E0ctKUrU6DZugCqABVnyADtDgBhZgGc4UYDCZwT6AGLsAGgFEAIgH1NAeQDKOufpQcGi6kjLyERjYEACuwQDa+ibwjCxoAMwADAA0aOLpALr+aAC+xUIiYuIATAIRFaIS4sWEYI01aAC0AHxVzRGt7ZI9Ic1lgsKN1dLFgUam5pbWtmgOLh7efgPkVDQ49Knorl4A6gByswajM5HymFixCUkpamg5eYXFZfWTVbXlvyaLQIbT+XV6QIGINEoXBfyI40GVQosJGTGqKPEQA), where you manually compute the positions of the ports and only use the `layered` algorithm?
